src/main/scala/util/AsyncQueue.scala

```diff
@@ -29,27 +29,38 @@ class AsyncQueueSource[T <: Data](gen: T, depth: Int, sync: Int) extends Module
   val bits = log2Ceil(depth)
   val io = new Bundle {
     // These come from the source domain
-    val enq  = Decoupled(gen).flip()
+    val enq  = Decoupled(gen).flip
     // These cross to the sink clock domain
     val ridx = UInt(INPUT,  width = bits+1)
     val widx = UInt(OUTPUT, width = bits+1)
     val mem  = Vec(depth, gen).asOutput
+    // Reset for the other side
+    val sink_reset_n = Bool().flip
   }
 
+  // extend the sink reset to a full cycle (assertion latency <= 1 cycle)
+  val catch_reset_n = AsyncResetReg(Bool(true), clock, !io.sink_reset_n)
+  // reset_n has a 1 cycle shorter path to ready than ridx does
+  val sink_reset_n = AsyncGrayCounter(catch_reset_n.asUInt, sync)(0)
```

Comment:
I suppose the fact that it's asynchronously reset is a major difference between the usual ShiftRegister. If we had a primitive that was "AsynchronouslyResetShiftRegister", would both of these things use it (not saying we should create such a monstrosity, just trying to understand what you're doing here).